src/main/java/com/google/sps/servlets/UserDataServlet.java

```diff
@@ -136,6 +196,22 @@ void doPostWrapper(DatastoreService datastoreService, HttpServletRequest request
     doPost(request, response);
   }
 
+  /** Wrap the doPost method for testing blobstore, for jUnit testing */
+  void doPostWrapper(DatastoreService datastoreService, BlobstoreService blobstoreService,
+                  HttpServletRequest request, HttpServletResponse response, boolean isTesting) throws IOException {
+    testing = isTesting;
+    blobstore = blobstoreService;
+    datastore = datastoreService;
+    doPost(request, response);
+  }
+
+  /** Wrap the getUploadedFileBlobKey method to test how Blobstore results are handled, for jUnit testing */
+  String getUploadedFileBlobKeyWrapper(HttpServletRequest request, String formInputElementName, BlobstoreService blobstoreService, boolean isTesting) {
+    blobstore = blobstoreService;
+    testing = isTesting;
+    return getUploadedFileBlobKey(request, formInputElementName);
+  }
```

Comment:
These methods, as well as the doGet/doPostWrapper methods preceding them, make it very difficult to reason about how behavior differs between tests/production.

In general, avoid putting any test-specific code alongside production code. If there is special setup that needs to be done for tests, it should happen in the test itself, and this class can expose whatever is necessary to accommodate that (e.g. make the `blobstore`/`datastore` fields package-private instead of private, and override them with mocks in your jUnit `#setUp` method).